### PR TITLE
Generate backward for upsample_linear1d.default from decomp.

### PR DESCRIPTION
**Pull request by amjames**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #122325
* __->__ #121769

### CODE REVIEW

**Comment by amjames** [comment]
I think this was missed in #114774 

The generated backward does not fuse with the forward, but the same can be seen for other `upsample_` ops, so I think this is okay.

**Comment by lezcano** [comment]
benchmarks?

**Comment by amjames** [comment]
Benchmarks here: 

Note: `Compile-pyimpl` uses a graph captured with python dispatch enabled, `Compile` will compile the forward pass, but uses the fallback for backward. 

```
[----------------------------------- upsample_linear1d.default -----------------------------------]
                                                            |  Compile-pyimpl  |  Compile  |  Eager
32 threads: ---------------------------------------------------------------------------------------
      input.shape=torch.Size([1, 3, 1000]), out_size=10     |        51        |     38    |   100 
      input.shape=torch.Size([1, 3, 1000]), out_size=100    |        62        |     53    |   100 
      input.shape=torch.Size([1, 3, 1000]), out_size=200    |        63        |     51    |   120 
      input.shape=torch.Size([1, 3, 20000]), out_size=10    |        61        |     49    |   100 
      input.shape=torch.Size([1, 3, 20000]), out_size=100   |        75        |     64    |   100 
      input.shape=torch.Size([1, 3, 20000]), out_size=200   |        75        |     64    |   100 
      input.shape=torch.Size([1, 3, 50000]), out_size=10    |        61        |     49    |   100 
      input.shape=torch.Size([1, 3, 50000]), out_size=100   |        75        |     64    |   200 
      input.shape=torch.Size([1, 3, 50000]), out_size=200   |        72        |     64    |   100 
      input.shape=torch.Size([5, 3, 1000]), out_size=10     |        49        |     52    |   100 
      input.shape=torch.Size([5, 3, 1000]), out_size=100    |        65        |     53    |   100 
      input.shape=torch.Size([5, 3, 1000]), out_size=200    |        64        |     53    |   100 
      input.shape=torch.Size([5, 3, 20000]), out_size=10    |        75        |     64    |   100 
      input.shape=torch.Size([5, 3, 20000]), out_size=100   |        75        |     64    |   100 
      input.shape=torch.Size([5, 3, 20000]), out_size=200   |        75        |     64    |   100 
      input.shape=torch.Size([5, 3, 50000]), out_size=10    |        76        |     64    |   120 
      input.shape=torch.Size([5, 3, 50000]), out_size=100   |        76        |     64    |   120 
      input.shape=torch.Size([5, 3, 50000]), out_size=200   |        76        |     64    |   120 
      input.shape=torch.Size([10, 3, 1000]), out_size=10    |        65        |     53    |   100 
      input.shape=torch.Size([10, 3, 1000]), out_size=100   |        65        |     53    |   100 
      input.shape=torch.Size([10, 3, 1000]), out_size=200   |        65        |     53    |   200 
      input.shape=torch.Size([10, 3, 20000]), out_size=10   |        76        |     64    |   120 
      input.shape=torch.Size([10, 3, 20000]), out_size=100  |        76        |     64    |   100 
      input.shape=torch.Size([10, 3, 20000]), out_size=200  |        76        |     64    |   100 
      input.shape=torch.Size([10, 3, 50000]), out_size=10   |       105        |     68    |   128 
      input.shape=torch.Size([10, 3, 50000]), out_size=100  |       107        |     70    |   131 
      input.shape=torch.Size([10, 3, 50000]), out_size=200  |       106        |     70    |   200 

Times are in microseconds (us).
```

@isuruf and I spoke a bit about this and it was not missed ,but intentionally left off because additional kernels are generated. 

I can see that the code generated by the compiler without python dispatch is actually *only* the fallback for backward being called. The compiler can see that the backward is only dependent on meta-data form the forward inputs and the `grad_out` param. I might not be accurate to say that the backward generated here is bad.

**Comment by lezcano** [comment]
Eager numbers look off there. Compile is falling back to eager, so those numbers can't be right. Perhaps it's not using cudagraphs.

A reasonable benchmark would be to benchmark forward + a reduction + backward, for example.

Also, those numbers are too small, use large numbers where you get times closer to the 1ms, for example. Also, use powers of 2 for batches and sizes, which are going to be more like those used in practice.

**Comment by isuruf** [comment]
Just to clarify, running
```python
import torch
aten = torch.ops.aten
dim = 1
x = torch.randn((4, 4, *[128]*dim), requires_grad=True, device='cuda')

def fn(x):
    if dim == 3:
        return aten.upsample_trilinear3d(x, [200, 200, 200], align_corners=True)
    elif dim == 2:
        return aten.upsample_bilinear2d(x, [200, 200], align_corners=True)
    else:
        return aten.upsample_linear1d(x, [200], align_corners=True)

y = torch.compile(fn)(x).sum()
y.backward()
```
generates 1 triton kernel for the forward pass and 3 triton kernels for the backward pass. I didn't time it, but I was thinking that 3 triton kernels for the backward would be slower.

**Comment by amjames** [comment]
Here is an updated benchmark: 

```
[----------------------------------- upsample_linear1d.default ------------------------------------]
                                                             |  Compile-pyimpl  |  Compile  |  Eager
32 threads: ----------------------------------------------------------------------------------------
      input.shape=torch.Size([64, 3, 16384]), out_size=512   |        172       |    153    |    362
      input.shape=torch.Size([64, 3, 16384]), out_size=128   |        163       |    138    |    342
      input.shape=torch.Size([64, 3, 16384]), out_size=64    |        162       |    135    |    340
      input.shape=torch.Size([64, 3, 32786]), out_size=512   |        348       |    241    |    572
      input.shape=torch.Size([64, 3, 32786]), out_size=128   |        338       |    221    |    551
      input.shape=torch.Size([64, 3, 32786]), out_size=64    |        336       |    219    |    548
      input.shape=torch.Size([128, 3, 16384]), out_size=512  |        343       |    308    |    701
      input.shape=torch.Size([128, 3, 16384]), out_size=128  |        326       |    280    |    663
      input.shape=torch.Size([128, 3, 16384]), out_size=64   |        321       |    274    |    660
      input.shape=torch.Size([128, 3, 32786]), out_size=512  |        688       |    478    |   1118
      input.shape=torch.Size([128, 3, 32786]), out_size=128  |        666       |    446    |   1076
      input.shape=torch.Size([128, 3, 32786]), out_size=64   |        674       |    442    |   1072
      input.shape=torch.Size([256, 3, 16384]), out_size=512  |        685       |    615    |   1385
      input.shape=torch.Size([256, 3, 16384]), out_size=128  |        657       |    555    |   1305
      input.shape=torch.Size([256, 3, 16384]), out_size=64   |        645       |    546    |   1298
      input.shape=torch.Size([256, 3, 32786]), out_size=512  |       1373       |    950    |   2217
      input.shape=torch.Size([256, 3, 32786]), out_size=128  |       1333       |    880    |   2129
      input.shape=torch.Size([256, 3, 32786]), out_size=64   |       1328       |    871    |   2122

Times are in microseconds (us).
```

For the benchmarks I am timing this:

```python
def f(args):
    input, out_sz, phony_grad = args
    out = torch.ops.aten.upsample_linear1d(
            input,
            [out_sz, ],
            align_corners=False,
            scales=0.6,
        )
    out.backward(phony_grad, retain_graph=True)
    return input.grad.sum(dim=-1) + out.sum(dim=-1)
```

I am seeing 2 kernels generated + a call to the fallback backward for compile without python dispatch enabled. 

With python dispatch enabled there are 3 kernels generated, one of which is just zero initializing buffers (and it is called twice).

FWIW I have tried this with `upsample_bicubic2d` and I see a similar pattern. With python dispatch enabled there are two additional kernels generated, one of which fills a buffer with zeros. However an important difference, in this case the zero initialization kernel is called a total of 16 times.

**Comment by isuruf** [comment]
Oh. That's an interesting optimization we don't have in inductor yet.
Going from
```
    %full_default : [num_users=2] = call_function[target=torch.ops.aten.full.default](args = ([4, 4, 128], 0), kwargs = {dtype: torch.float32, layout: torch.strided, device: cuda:0, pin_memory: False})
    %_unsafe_index_put : [num_users=1] = call_function[target=torch.ops.aten._unsafe_index_put.default](args = (%full_default, [None, None, %clamp_max], %mul_2, True), kwargs = {})
    %_unsafe_index_put_1 : [num_users=1] = call_function[target=torch.ops.prims._unsafe_index_put_.default](args = (%full_default, [None, None, %convert_element_type_1], %add_2, True), kwargs = {})
    %add_3 : [num_users=1] = call_function[target=torch.ops.aten.add.Tensor](args = (%_unsafe_index_put, %_unsafe_index_put_1), kwargs = {})
```
to
```
    %full_default : [num_users=2] = call_function[target=torch.ops.aten.full.default](args = ([4, 4, 128], 0), kwargs = {dtype: torch.float32, layout: torch.strided, device: cuda:0, pin_memory: False})
    %_unsafe_index_put : [num_users=1] = call_function[target=torch.ops.aten._unsafe_index_put.default](args = (%full_default, [None, None, %clamp_max], %mul_2, True), kwargs = {})
    %add_3 : [num_users=1] = call_function[target=torch.ops.prims._unsafe_index_put_.default](args = (%_unsafe_index_put, [None, None, %convert_element_type_1], %add_2, True), kwargs = {})
```

**Comment by lezcano** [comment]
I don't really understand why do we have an `aten._unsafe_index_put` and a `prims._unsafe_index_put`...

**Comment by lezcano** [comment]
>With python dispatch enabled there are 3 kernels generated, one of which is just zero initializing buffers (and it is called twice).

torch.compile is always run with PyTorch dispatch enabled, so this is how benchmarks should be run if you don't use torch.compile directly.

**Comment by amjames** [comment]
> torch.compile is always run with PyTorch dispatch enabled, so this is how benchmarks should be run if you don't use torch.compile directly.

Right, I am just differentiating between the results which use the change here and the results which use the fallback for the backward. 

FWIW: I don't see any calls to `prims._unsafe_index_put_`  just `aten._unsafe_index_put` being called twice.

**Comment by isuruf** [comment]
The two zero kernel calls can be reduced down to one by using https://github.com/isuruf/pytorch/commit/4c409c124a023683413165152d8a29a3864c6165#diff-5d0c4891ce10ae399cd6d44f7408532d1a9eb56103ab0cd7db3613cdf4fbaf50R358-R387

**Comment by lezcano** [comment]
Very cool FX graph pass, @isuruf.  Can you submit a separate PR?

**Comment by ezyang** [comment]
Even if the performance is worse, generating the backward in this way is still useful because it is deterministic (and the eager kernel usually is not)

**Comment by ezyang** [comment]
cc @bhack

**Comment by bhack** [comment]
@ezyang I don't know this part of the code. What is this single line change doing?

**Comment by lezcano** [comment]
Where does the non-determinism come from in eager? This is not fully deterministic either, as the backward of `index` uses atomic adds.

**Comment by ezyang** [comment]
@lezcano we have a deterministic implementation of index backwards though, right?

The nondeterminism in eager upsample is due to atomicAdds when we accumulating gradients from repeated uses of the same input location (since this is upsampling)

**Comment by lezcano** [comment]
index_backwards is just `index_put(accumulate=True)`, which is implemented via atomic adds in triton, as well as in eager mode
https://github.com/pytorch/pytorch/blob/b524a404e0ab6de43cbbb27a5a72dc93a8c24d3b/torch/_inductor/lowering.py#L3257

**Comment by isuruf** [comment]
@lezcano, `index_put` does not reach that statement when deterministic is enabled, right? See https://github.com/pytorch/pytorch/blob/b524a404e0ab6de43cbbb27a5a72dc93a8c24d3b/torch/_inductor/lowering.py#L3193-L3195

**Comment by lezcano** [comment]
Ah, right, we do have an eager deterministic implementation via sort. That being said, we cannot implement that on triton, so the point above still stands.

**Comment by bhack** [comment]
> Ah, right, we do have an eager deterministic implementation via sort. That being said, we cannot implement that on triton, so the point above still stands.

What do you mean? If we are using `torch.are_deterministic_algorithms_enabled():` we are going to break the graph and run a eager node with a fallback sort?

**Comment by lezcano** [comment]
yes, `are_deterministic_algorithms_enabled` trade performance for determinism. In this case, we don't have a native deterministic `index_put` implementation, since triton does not currently implement sorting.

**Comment by bhack** [comment]
Is it deterministic also in this case?
https://github.com/pytorch/pytorch/blob/72662bf05b3499ce96aae9183a489c78f0c44c84/torch/_inductor/lowering.py#L3178-L3180

Why we don't have also `float16` in the list as I don't see a `float16` kernel in triton:
https://github.com/openai/triton/blob/main/test/Conversion/tritongpu_to_llvm.mlir#L1068-L1126

**Comment by lezcano** [comment]
Whenever there is a fallback and the deterministic flag is on, we generate a deterministic algorithm. You can try it yourself executing that code and looking at the generated code and the kernels that are executed.

**Comment by bhack** [comment]
But why we don't have `float16` in that list and only `bfloat16`?

**Comment by amjames** [comment]
@lezcano / @ezyang since #122415 did not work out, is there any need for this? 

@isuruf has #123222 which decomposes the backward directly

**Comment by lezcano** [comment]
We can close this one. Thank you for the exploratory effort tho.
